src/mono/mono/mini/interp/transform-opt.c

```diff
@@ -481,45 +484,1072 @@ interp_alloc_offsets (TransformData *td)
 	}
 	final_total_locals_size = ALIGN_TO (final_total_locals_size, MINT_STACK_ALIGNMENT);
 
-	// Iterate over all call args locals, update their final offset (aka add td->total_locals_size to them)
-	// then also update td->total_locals_size to account for this space.
-	td->param_area_offset = final_total_locals_size;
+	// Iterate over all call args locals, update their final offset (aka add td->total_locals_size to them)
+	// then also update td->total_locals_size to account for this space.
+	td->param_area_offset = final_total_locals_size;
+	for (unsigned int i = 0; i < td->vars_size; i++) {
+		// These are allocated separately at the end of the stack
+		if (td->vars [i].call_args) {
+			td->vars [i].offset += td->param_area_offset;
+			final_total_locals_size = MAX (td->vars [i].offset + td->vars [i].size, final_total_locals_size);
+		}
+	}
+	td->total_locals_size = ALIGN_TO (final_total_locals_size, MINT_STACK_ALIGNMENT);
+}
+
+/*
+ * DOMINANCE COMPUTATION
+ */
+
+static GString*
+interp_get_bb_links (InterpBasicBlock *bb)
+{
+	GString *str = g_string_new ("");
+
+	if (bb->in_count) {
+		g_string_append_printf (str, "IN (%d", bb->in_bb [0]->index);
+		for (int i = 1; i < bb->in_count; i++)
+			g_string_append_printf (str, " %d", bb->in_bb [i]->index);
+		g_string_append_printf (str, "), ");
+	} else {
+		g_string_append_printf (str, "IN (nil), ");
+	}
+
+	if (bb->out_count) {
+		g_string_append_printf (str, "OUT (%d", bb->out_bb [0]->index);
+		for (int i = 1; i < bb->out_count; i++)
+			g_string_append_printf (str, " %d", bb->out_bb [i]->index);
+		g_string_append_printf (str, ")");
+	} else {
+		g_string_append_printf (str, "OUT (nil)");
+	}
+
+	return str;
+}
+
+static int
+dfs_visit (TransformData *td)
+{
+	int dfs_index = 0;
+	int next_stack_index = 0;
+	td->bblocks = (InterpBasicBlock**)mono_mempool_alloc0 (td->opt_mempool, sizeof (InterpBasicBlock*) * td->bb_count);
+	InterpBasicBlock **stack = (InterpBasicBlock**)g_malloc0 (sizeof (InterpBasicBlock*) * td->bb_count);
+
+	g_assert (!td->entry_bb->in_count);
+	stack [next_stack_index++] = td->entry_bb;
+
+	while (next_stack_index > 0) {
+		// Pop last added element
+		next_stack_index--;
+		InterpBasicBlock *bb = stack [next_stack_index];
+
+		// Process current bblock
+		td->bblocks [dfs_index] = bb;
+		bb->dfs_index = dfs_index++;
+
+		// Push all nodes to process next
+		for (int i = 0; i < bb->out_count; i++) {
+			InterpBasicBlock *out_bb = bb->out_bb [i];
+			if (out_bb->dfs_index == -1) {
+				stack [next_stack_index++] = out_bb;
+				// Mark node as gray so it is not pushed again
+				out_bb->dfs_index = -2;
+			}
+		}
+	}
+
+	g_free (stack);
+	return dfs_index;
+}
+
+static void
+interp_compute_dfs_indexes (TransformData *td)
+{
+	for (InterpBasicBlock *bb = td->entry_bb; bb != NULL; bb = bb->next_bb)
+		bb->dfs_index = -1;
+	// Sort bblocks in reverse postorder
+	int dfs_index = dfs_visit (td);
+	td->bblocks_count = dfs_index;
```

Comment:
it might be nice to name this one `bblocks_count_no_eh` so the difference is obvious when reading code, otherwise it's a potential landmine for people who aren't super familiar with the nuances of this system